### ISSUE TITLE
Run join benchmark with UCX

### ISSUE THREAD

**Issue by mrocklin** [rapidsai/dask-cudf#287]
We should test dask-cudf join performance with UCX.  Today this involves using a few in-development branches:

-  UCX: As described in [this post](https://blog.dask.org/2019/06/09/ucx-dgx) there is now a mostly working UCX + UCX-Py + Dask integration.  @jakirkham has conda packages up for it here:

    ```
    conda create -n ucx -c conda-forge -c jakirkham/label/ucx cudatoolkit=9.2 ucx-proc=*=gpu ucx ucx-py python=3.7
    ```

    https://github.com/jakirkham/staged-recipes/tree/17005b662e392672de7a82778b07eb4dec8b5ad9/recipes
-  cuDF serialization support, as improved by @rjzamora here: https://github.com/rapidsai/cudf/pull/1947

-  There is an old notebook doing this [here](https://nbviewer.jupyter.org/urls/gist.githubusercontent.com/mrocklin/6e2c33c33b32bc324e3965212f202f66/raw/6ab6a32c90391e1a54b2a335266443d9900b56fb/dask-cudf-join.ipynb) although we would want to make a few changes:
    - Use the `DGX` class from dask_cuda dev version, probably with only half of the GPUs for now, so something like `cluster = DGX(cuda_visible_devices=[0,1,2,3])`
    - We might want to fool around with the data sizes a bit to make sure that things fit in only half of the device memory.

cc @rjzamora and @madsbk , both of whom I think might be interested in this

**Comment by mrocklin** [comment]
Also cc @randerzander and @kkraus14 who might want to track this issue

**Comment by nsakharnykh** [comment]
cc @gaohao95

**Comment by quasiben** [comment]
I created a gist with the current notebook and task stream profile here:
https://gist.github.com/quasiben/2530898f923d3cddcd34c6bed2a88543

Also attaching a screen shot of the task stream

<img width="508" alt="Screen Shot 2019-07-08 at 2 49 48 PM" src="https://user-images.githubusercontent.com/1403768/60835028-259e3a80-a190-11e9-833d-bfcf00cf9774.png">

As you can see there isn't much communication visualized.  This was run on a DGX box with v100s (32GB RAM).   The join primarily stayed ~16GB but one GPUs approached the max:

```
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0     33902      C   ...len/miniconda3/envs/cudf-dev/bin/python 11090MiB |
|    1     33900      C   ...len/miniconda3/envs/cudf-dev/bin/python 30022MiB |
|    2     33904      C   ...len/miniconda3/envs/cudf-dev/bin/python 12050MiB |
|    3     33898      C   ...len/miniconda3/envs/cudf-dev/bin/python 13130MiB |
+-----------------------------------------------------------------------------+
```

I am going to rerun with less data

**Comment by kkraus14** [comment]
@quasiben It may be worth passing as_index=False to the merge command as well as I'm not sure if we're handling serialization of the Indexes properly.

**Comment by quasiben** [comment]
In the middle of a debug session with @rjzamora I noticed that there were issues deserializing the distributed cuDF object.  I have since filed an issue here:
https://github.com/dask/distributed/issues/2830

**Comment by quasiben** [comment]
@kkraus14 there is no option fo `as_index=False` for `merge` ?

**Comment by kkraus14** [comment]
@quasiben I had a brainfart and mixed up groupby and join, disregard 😄

**Comment by mrocklin** [comment]
Things should change when the second dataframe has more than one partition.  I recommend playing around a bit with relative sizes.

**Comment by mrocklin** [comment]
To be explicit here, I think that a success here would be a task stream that showed many red bars, but hopefully those bars would be small and correspond to multi-GB/s transfers.

**Comment by quasiben** [comment]
@mrocklin, indeed, with second dataframe having more than one partition more communication does occur.  However, so does deserialization and errors like following arise:

> File "/home/nfs/bzaitlen/GitRepos/cudf/python/cudf/cudf/dataframe/columnops.py", line 36, in __init__ super(TypedColumnBase, self).__init__(**kwargs) File "/home/nfs/bzaitlen/GitRepos/cudf/python/cudf/cudf/dataframe/column.py", line 160, in __init__ self._update_null_count(null_count) File "/home/nfs/bzaitlen/GitRepos/cudf/python/cudf/cudf/dataframe/column.py", line 173, in _update_null_count assert 0 <= null_count <= len(self) TypeError: 'NoneType' object cannot be interpreted as an integer

I've already logged this issue here: https://github.com/dask/distributed/issues/2830 and I'm going to be looking at it more closely now
